docs: add localize migration doc (#33086)
angular/angular: aio/content/guide/migration-localize.md

# `$localize` Global Import Migration

## Why is this migration necessary?

> 

Reviewer: Proposed revision for this section
````markdown
Angular compiles HTML templates into generated TypeScript code. Prior to Angular v9,
Angular's internationalization (i18n) system inlined translated messages into the compiled
output as part of this template compilation. This approach required running the template
compiler once per target locale, often leading to slow production build times.

The Angular Ivy i18n system instead performs template compilation exactly once. Then,
as a _post-compilation_ step, Angular inlines messages for each target locale. By separating
this message inlining from template compilation, the process can run much faster.

However, in order for this new approach to work, each application
must now import new code from the `@angular/localize` package. If
this import is missing, you will see an error message like this:

```
Error: The global function `$localize` is missing.
Please add `import '@angular/localize/init';` to your polyfills.ts file.
```

This schematic automatically adds the `@angular/localize` import for you
if your app uses Angular's i18n APIs.
````

Author: Beautiful, thank you! I didn't know how to articulate some of what I wanted to say and this really helps.

To keep Ivy fast, inlining in Angular version 9 happens after build. After you have the minified bundle, all the strings are inlined, so your app is compiled one time, regardless of how many languages you're translating.

Apps running versions of Angular prior to 9 will return a runtime error when trying to render a template that contains i18n or related attributes. For example:

```shell
Error: The global function `$localize` is missing.
Please add `import '@angular/localize/init';` to your polyfills.ts file.
```

This schematic migrates your app so that you won't get this error.

## What does this schematic do?

If you're using i18n, this schematic imports `@angular/localize` to `polyfills.ts` and adds it to `package.json`.

```ts
/******************************************************************
 * Load `$localize` - used if i18n tags appear in Angular templates.
 */
import '@angular/localize/init';
```

> `@angular/localize` is a new package that works with Ivy. The `$localize` function looks up the translation in your translation file and inlines translated strings.

Reviewer: I think it would be valuable to explain here that this new code doesn't add any run-time overhead (I believe, Kara can confirm).

Author: Added and leaving for @Kara to confirm.

## Should I update my library?

> You don't need to update your libraries because libraries are still shipping with View Engine.

Reviewer: I feel like this section needs more elaboration, but I don't know enough to propose anything

Author: Maybe @kara has some thoughts?

Author: Maybe @Kara has some thoughts?

# `$localize` Global Import Migration

## Why is this migration necessary?

> Angular compiles HTML templates into generated TypeScript code. Prior to Angular version 9,

Reviewer 2: ```suggestion
Prior to Angular version 9,
```

Angular's internationalization (i18n) system inlined translated messages into the compiled

> output as part of this template compilation. This approach required running the template

Reviewer 2: ```suggestion
output as part of the CLI build process. This approach required running the template
```

Reviewer 3: The new approach in v9 can also inline the translated messages as part of the CLI build process. So I think this needs to be more nuanced than that.

Reviewer 3: I think that talking about the Angular compiler here is correct. The CLI is a layer above this.

Reviewer 2: Ok let's keep it as is

compiler once per target locale, often leading to slow production build times.

> The Angular Ivy i18n system instead performs template compilation exactly once. Then,

Reviewer 3: I would say something along the lines of:

>>>
In the new i18n system, the Angular compiler tags i18n messages in the compiled code with a global `$localize` handler. The inlining of translations then occurs as a post-compilation step for each locale. Because the application does not need to be built again for each locale, this makes the process much faster.

Reviewer 3: By the way, should we be avoiding the use of the word `ivy` in these docs?

Author: Added your change, @petebacondarwin. Regarding the use of the word `Ivy`, I hadn't thought of that, but it makes sense. What should I use as an alternative?

Reviewer 3: Let's ask @kara :-)

as a _post-compilation_ step, Angular inlines messages for each target locale. By separating
this message inlining from template compilation, the process can run much faster.

However, in order for this new approach to work, each application
must now import new code from the `@angular/localize` package. If

> this import is missing, you will see an error message like this:

Reviewer 3: And then here you could say:

>>>
The post-compilation inlining step is optional - for example during development or if the translations will be inlined at runtime. Therefore this global `$localize` must be available on the global scope at runtime.
To make `$localize` available on the global scope, each application must now import the `@angular/localize/init` module. This has the side-effect of attaching a minimal implementation of `$localize` to the global scope.

If this import is missing, you will see an error message like this:

```
Error: The global function `$localize` is missing.
Please add `import '@angular/localize/init';` to your polyfills.ts file.
```

> This schematic automatically adds the `@angular/localize` import for you

Reviewer 3: ```suggestion
This schematic automatically adds the `@angular/localize/init` import for you
```

if your app uses Angular's i18n APIs.

> ## What does this schematic do?

Reviewer 2: This should always be the first question. Move up to the top?

> If you're using i18n, this schematic imports `@angular/localize` to `polyfills.ts` and adds it to `package.json`.

Reviewer 2: Let's add a diff for package.json?

```suggestion
If you're using i18n, this schematic adds an import statement for `@angular/localize` to `polyfills.ts`. It will look something like this:

<example diff for polyfills.ts>

It also lists `@angular/localize` as a dependency in your app's `package.json` to ensure the import is found.

<example diff for package.json>
```

Reviewer 2: I think we're also missing the `no-import-side-effect` change in tslint.json

Reviewer 3: We didn't add the tslint change in the end because the standard CLI project does not include the rule (`no-import-side-effect`) that needs to be overridden.

Reviewer 3: But if one has it enabled manually then perhaps it is useful to mention it here?

Reviewer 2: Ah, didn't realize. We should probably add an faq like "why is my tslint failing?" In case people have a custom config

Author: Added:

>## Why is my tslint failing?

The schematic updates the `tslint.json` config file by adding the `@angular/localize` path to the `no-import-side-effect` rule if you have enabled it manually.

```json

"no-import-side-effect": [
  true,
  {
    "ignore-module": "(core-js/.*|zone\\.js/.*|@angular\/localize)$"
  }
]

```

Should I say something about changing it? Because if it's failing, and the reason is because of `no-import-side-effect` being set to`true`, does that mean one should change it to `false`? @petebacondarwin @kara

Reviewer 2: As @petebacondarwin mentioned above, the schematic no longer updates the tslint. You should say something like:

> The import of `@angular/localize/init` may cause a tslint error for 'no-import-side-effect' because it adds to the global context (i.e. a side effect). To fix this error, add the following to your tslint.config:

<diff>

```ts
/******************************************************************
 * Load `$localize` - used if i18n tags appear in Angular templates.
 */
import '@angular/localize/init';
```

> `@angular/localize` is a new package that works with Ivy that doesn't add any run-time overhead. The `$localize` function looks up the translation in your translation file and inlines translated strings.

Reviewer 2: For v9, we are just doing compile-time inlining, so the $localize function isn't actually being called. The i18n inliner looks for`$localize` function calls and replaces the strings inside them with the proper translations (at build-time).  We just need the `$localize` function to be defined globally so the app will build (because the inlining happens after build).

When we do actual runtime i18n (probably v10), $localize will be a real runtime function that loads translations on the fly, but not yet. You can see more info here: https://github.com/angular/angular/blob/master/packages/localize/init/index.ts#L92

```suggestion
`@angular/localize` is a new package that works with Ivy that doesn't add any run-time overhead. It simply adds a `$localize` symbol to the global context, which allows i18n to work properly.
```

Reviewer 3: It is a bit more subtle... when developing (or just if you happen to have i18n tags but no translations) then the `$localize` calls do end up at the browser and do get called. The `@angular/localize/init` provides a minimal version of `$localize` that strips out any metadata and just renders the untranslated strings.

Reviewer 2: Ahh, right, I forgot about dev mode. Let's replace "symbol" with "function"and that should encompass both

Author: Have edited it. Am I making sense, @petebacondarwin?

>`@angular/localize` is a new package that works with Ivy that doesn't add any run-time overhead.
It adds a `$localize` symbol to the global context, which allows i18n to work properly.
Specifically, when developing (or just if you happen to have `i18n` tags but no translations), the `$localize` are called.
The `@angular/localize/init` provides a minimal version of `$localize` that strips out any metadata and just renders the untranslated strings.

Reviewer 2: I don't think we need to talk about how $localize works in development mode.  Can we just replace "symbol" with "function"?

> @angular/localize is a new package that works with Ivy that doesn't add any run-time overhead. It adds a $localize function to the global context, which allows i18n to work properly.

Reviewer 3: It makes sense but I am afraid it will confuse people who perhaps don't get the full picture....

Let's not mention overhead. It will just make people worry that it actually is adding overhead and we are trying to hide something :-)

How about something like:

> `@angular/localize` is a new package that supports i18n of messages in ivy(?) applications.
> It works by tagging such messages with a `$localize` symbol in the compiled code.
> The CLI can translate these tagged messages, inlining the translations in place of the tagged message, leaving no trace of `$localize`.
> During initial development (or if not doing compile time inlining of translations) a minimal implementation of `$localize` is needed and render tagged messages, stripping any metadata.
> This implementation must exist as a function on the global scope (e.g. the `Window`).
> This minimal implementation is loaded by importing the `@angular/localize/init` module, which has the side-effect of attaching the `$localize` function to the global scope.

Reviewer 2: Fair enough. I just worry about overwhelming people with too much detail, but it's a balance. Your suggestion works for me.

Reviewer: IMO, the _migration_ docs should go into an absolute bare minimum of technical detail. The actual docs for the feature should be the canonical source of the full technical details.

Reviewer 3: Actually I agree. So something like... ?

> `@angular/localize` is a new package that supports i18n of messages in ivy(?) applications.
> This package requires a global `$localize` symbol to exist.
> The symbol is loaded by importing the `@angular/localize/init` module, which has the side-effect of attaching it to the global scope.

Author: Thank you very much for this, everyone. Added Pete's latest^^. Will updated if it evolves further. :)

## Should I update my library?

> You don't need to update your libraries because libraries are still shipping with View Engine.

Reviewer 2: I think we can actually drop this section since 99% of libraries don't use i18n

# `$localize` Global Import Migration

## What does this schematic do?

If you're using i18n, this schematic adds an import statement for `@angular/localize` to `polyfills.ts` that will look something like this:

```ts
/******************************************************************
 * Load `$localize` - used if i18n tags appear in Angular templates.
 */
import '@angular/localize/init';
```

It also lists `@angular/localize` as a dependency in your app's `package.json` to ensure the import is found.

```json

"dependencies": {
  ...
  "@angular/localize": "...",
  ...
}

```

`@angular/localize` is a new package that supports i18n of messages in Ivy applications.
This package requires a global `$localize` symbol to exist.
The symbol is loaded by importing the `@angular/localize/init` module, which has the side-effect of attaching it to the global scope.

## Why is this migration necessary?

Prior to Angular version 9, Angular's internationalization (i18n) system inlined translated messages into the compiled output as part of this template compilation. This approach required running the template compiler once per target locale, often leading to slow production build times.

> In the new i18n system, the Angular compiler tags i18n messages in the compiled code with a global $localize handler.

Reviewer 3: ```suggestion
In the new i18n system, the Angular compiler tags i18n messages in the compiled code with a global `$localize` handler.
```

The inlining of translations then occurs as a post-compilation step for each locale.
Because the application does not need to be built again for each locale, this makes the process much faster.

The post-compilation inlining step is optional&mdash;for example during development or if the translations will be inlined at runtime.
Therefore this global $localize must be available on the global scope at runtime.

> To make $localize available on the global scope, each application must now import the @angular/localize/init module.

Reviewer 2: ```suggestion
To make $localize available on the global scope, each application must now import the `@angular/localize/init` module.
```
